### DISCUSSION TITLE
Stretch video with MediaEnginePlayer

### DISCUSSION THREAD

**Discussion by wsandor** [FactoryXCode/MfPack#84] [Q&A] [answered]
Hi,
Could anybody help me how to stretch the video output to the size of the output window with MediaEnginePlayer? By default the output is letter/pillarboxed. This task was easy with DirectShow or with the MfPlayer, but they are deprecated according to Microsoft, so I would like to use MediaEnginePlayer. Maybe the solution is VideoProcessorMFT/IMFVideoProcessorControl, but unfortunately I don't know how to insert it into a simple player like in the MfMediaEnginePlayer1 demo. Any help is apperciated.

**Comment by FactoryXCode** [comment]
Hi,

Please take a look in to the MfMediaEnginePlayer2 sample.
Regards Tony.

**Reply by wsandor** [reply]
Dear Tony,
Thanks for your kind reply!
Maybe I miss something, but unfortunately I can't find it there. I can set the aspect ratio there, but if it doesn't match the video's ar. then it letterboxes or pillarboxes the video (there will be black parts at the sides or the top/down part of the video. My aim is to the video fill entirely the player window, regardless of the aspect ratio - so distort the video if needed. Do you find it possible somehow?

**Reply by FactoryXCode** [reply, marked as answer]
Hi,
To adjust the video dimensions in IMFMediaEngineEx you should use IMFMediaEngineEx function UpdateVideoStream. I don't know if it is sufficient to do on what you are trying to do, but, as far as I do know, that's the only way to adjust video frames with this interface.
Regards Tony.

**Comment by wsandor** [comment]
Dear Tony,
Thanks again for your answer! Yes, I've tried UpdateVideoStream, but unfortunately it letter/pillarboxes the output if the AR of the video doesn't match the output windows's AR. I've read somewhere that this could be done with Video Resizer DSP or Video Processor MFT throught IMFMediaEngineEx.InsertVideoEffect but unfortunately I haven't succeeded with it yet.

**Comment by FactoryXCode** [comment]
Well, happy new year first.
As you want, there is an interesting article about this [here](https://learn.microsoft.com/en-us/windows/win32/medfound/video-processor-mft) Maybe this will help you further. However, this article is also pointing to the solution I suggested (IMFMediaEngineEx).
I doubt it will do the thingies you want, because on what I understand is that you actually want to stretch pixels to match a surface?
Regards, Tony.

**Comment by wsandor** [comment]
Dear Tony,
Thanks for your reply and the link, I will study it.
Yes, you understand it right, I want to stretch pixels to match a surface. (as it was easyly achieveable with DirectShow)

**Comment by FactoryXCode** [comment]
To manipulate the pixel height and width (Pixel Aspect Ratio or PAR), you could use the GetPixelAspectRatio, SetPixelAspectRatio, SetOutputRectangleAspectRatio etc. methods declared in the WinApi.MediaFoundationApi.MfMetLib unit.
This unit contains lots of methods usefull for Media Foundation.
See also [here](https://learn.microsoft.com/en-us/windows/win32/medfound/picture-aspect-ratio) for explanations.

Regards, Tony.
